java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java

```diff
@@ -31,10 +32,12 @@
 import org.sonar.plugins.java.api.tree.VariableTree;
 
 import java.util.List;
+import java.util.function.Predicate;
 
 @Rule(key = "S2039")
 public class FieldModifierCheck extends IssuableSubscriptionVisitor {
 
+  private static final Predicate<ModifiersTree> MODIFIERS_TREE_PREDICATE = new VisibleForTestingPredicate().negate();
```

Comment:
Tests of rules are not integration tests, but unit test. Each rule is tested unitarilly, with expected behavior. You don't need to try every possible combination of modifiers, but only the presence or not of the annotation.

I can however tell you that mocking absolutely everything which is generated by the parsing is not a better approach in order to test the behavior of your predicate (which could be a simple lambda expression on a stream of ModifierTree).